### ISSUE TITLE
ROCm wvSplitK default drifts the Qwen3.5-0.8B sacred anchor on gfx1100

### ISSUE THREAD

**Issue by VikashLoomba** [mudler/vllm.cpp#2772]
Row: `BACKEND-ROCM`

## Recorded defect

Current `main` at `4d10c8acc527c34a6a58a309d52ea5f8fbd1d47b` fails the dedicated Qwen3.5-0.8B gfx1100 sacred gate:

```text
anchor drift prompt[10] tok=10 engine=369 committed anchor=488
```

The binary ran the only doctest case, selected native ROCm providers, and reported no reference-tier provider. This is not a zero-test or missing-artifact result.

## Causal evidence

A GPU-locked bisect rebuilt and ran the real gate at every candidate between known-good `7b89cf30775df79155d4bcc4f3cabf57637f74fc` and known-bad `fb7ff54870bdbeeaa94338a4b2c73dee1c1efaed`.

The first bad commit is:

```text
f38c1edc4ea679348f856c0c0b20fb0702f77daf
perf(BACKEND-ROCM): route decode-skinny GEMMs to a ported wvSplitK kernel (#487, #1183) (#506)
```

The known-good source rebuilt under the same current HIP 7.15 toolchain and passed 1/1 case and 137/137 assertions. Current `main` with `VT_ROCM_SKINNY=0` also passed 1/1 case and 137/137 assertions. The default skinny arm failed. Disabling cooperative Q8_K or D128 decode attention did not change the failure.

`VT_DUMP_IDS=1` shows six changed cells, all in prompt 10 from token 10 onward. Only token 10 shares the committed input prefix, so the old teacher-forced gaps cannot adjudicate later suffix cells.

## Scope

1. Reproduce the default-on skinny result and the BLAS rollback on gfx1100 using the pinned Qwen3.5-0.8B checkpoint.
2. Capture the exact skinny MatmulBT shapes reached before the first divergent token.
3. Compare the skinny and BLAS outputs at those shapes with the pinned upstream kernel behavior and the pinned vLLM oracle on the identical prefix.
4. Decide from oracle evidence whether to repair the kernel arithmetic, narrow its default eligibility, or rederive the model golden because the port intentionally mirrors the upstream production path.
5. Add a focused regression that enters through `LoadedEngine::FromModelDir` and fails if the selected resolution is removed.
6. Mutate the production skinny call site and prove the focused gate detects the change.

## Boundaries

- Do not refresh the committed golden before the pinned oracle adjudicates the first changed token on the identical prefix.
- Do not classify the new token as correct or incorrect from the old gap table after prefix divergence.
- Do not accept throughput evidence while the default sacred gate is red.
- Issue #2773, which supersedes #1588, owns CPU/ROCm boundary and state characterization. This issue owns the skinny-GEMM sacred-gate repair.

## Current position

PR #2856 carries the reviewed repair at `f544a5d5784178a843184028edf3011be62dc28f`. Production-config vLLM at historical pin `5559679229bc961848b121ccdeaa8fa5d79bec98` adjudicated the first disputed token as 369. The repair defaults the capture tools to production mode and refreshes the evidence; it changes no kernel arithmetic or acceptance threshold.

The operator reran the byte-identical integration build at `f06619e4c`: CPU script modes 2/2, prerequisites 77/77/77/86, focused wvSplitK 79,796 assertions, and sacred 137 assertions with 16 accepted prompts, 15 strict, one tied, maximum gap zero, and zero forward divergence. The final head adds only the independently reviewed config SHA-256 correction. Its controlled preflight exited 0 with ten explicit pending skips before the exact-SHA push.

Landing remains PENDING because the contributor account has no upstream merge permission. The tested integration base is `abc4be38f55134811a96d6aa3fd99e717e69cb82`; this evidence does not validate newer main changes or active vLLM `e126687a9a828d513c01a07cd69f025f27d63280`. Issue #2773 owns this model's cache-matched active-pin captures.

**Comment by VikashLoomba** [comment]
Pinned-oracle adjudication of the first shared-prefix divergence is now complete.

The restored oracle image uses:

- vLLM source `5559679229bc961848b121ccdeaa8fa5d79bec98`
- wheel SHA-256 `c4892c729daaff1ac3e8ad72d4e4753118b84e7d7c265634a1326b261a44ee43`
- immutable `rocm/vllm-dev` base digest `sha256:ba82bf93fc2d3787550b77fb2a79fe230130f05bec737c671093251c55459309`
- image ID `sha256:80aab4c182a1f3eeebe286173977e57fcaf10a049b41f475655b35d285de31dc`
- ROCm userland 7.2.53211 and the host gfx1100 device

The GPU smoke and model run both held `/home/vikash/gpu.lock`. The model run used the local `Qwen/Qwen3.5-0.8B@2fc06364715b967f1860aea9cf38778875588b17` snapshot with network access disabled. The runtime resolved production configuration, not eager fallback: `enforce_eager=False`, `CUDAGraphMode.FULL_AND_PIECEWISE`. `VLLM_ROCM_USE_SKINNY_GEMM` was unset in the environment and resolved to `True` in the engine configuration.

For prompt 10, `The mitochondria is the powerhouse of`, the oracle tokenizer produced the exact seven stored prompt IDs. The current local default/wvSplitK 16-token sequence matched the pinned oracle's production greedy sequence exactly:

```text
279 2691 13 271 248068 271 248069 271 760 4952 369 2972 1802 159034 271 760
```

At the first disputed cell, token 10 on that identical prefix, teacher-forced pinned-vLLM log probabilities were:

```text
369  -1.2000471353530884
488  -1.2000471353530884
```

The tokens are exactly tied at the precision returned by the oracle. The oracle's greedy result and teacher-forced argmax are both 369. Thus the current local default follows the pinned upstream default at the first shared-prefix cell; this evidence does not support treating token 488 as the uniquely correct result or disabling wvSplitK merely to preserve the old local anchor.

Evidence is retained under `/tmp/qwen35-skinny-oracle-prefix-2772-production`: `result.json` SHA-256 `05eb66ca1e4a3d579a27bf43996200a53c316b97374bd8d183420aef7a0412df`, stdout `b92f015fc09a034780ff22262016e34c4ab9ef97c5764fabb4ff228240800f27`, and stderr `8c4745aa687f54750d7ea75036b55ac3671b4de1fcf258962d949ebbd93aaece`.

I am running the production-config teacher-forced check over all 256 current local cells before selecting the repair. The committed anchor and gap artifacts will not change unless that complete check is green and the focused regression/mutation detects the default production route.

**Comment by VikashLoomba** [comment]
The complete production-config teacher-forced check is green and selects the oracle-supported re-derivation branch.

The run used the same image, local checkpoint, mutex, device, and network-disabled configuration recorded above. It compared the current default local 16-prompt × 16-token dump against pinned-vLLM logits on each exact local prefix. Results:

- prompt-tokenization mismatches: 0/16
- local token outside oracle top 20: 0/256
- nonzero teacher-forced gaps: 0/256
- strict oracle-argmax matches: 256/256
- maximum gap: 0.0 nats

Only six local anchor cells change, all in prompt 10 after the shared-prefix split: positions 10 through 15. The current sequence is oracle-supported at every resulting suffix position. The regenerated gap array also changes two cells, `(7,8)` and `(10,10)`, from the historical 125 milli-nats to 0 under the restored production-config oracle.

Evidence is retained under `/tmp/qwen35-skinny-oracle-full-gap-2772-production`:

- `result.json`: `495ff7624c7a46fd135edc18fddc364ef449a10db78a5eb3e5b7d2fc1a99589b`
- current local anchor candidate: `395ecea93be0c2cf036a0d8d515493a5178a5dcaea172df8ff71cfd80ab24112`
- gap candidate: `c707d168d23aea394987c4a40c4e92d8347c0400e9790df0851c94d37c113a82`
- teacher-forced argmax array: `395ecea93be0c2cf036a0d8d515493a5178a5dcaea172df8ff71cfd80ab24112`
- stdout: `3a517aea0b4b7716b99a06b0ef425375e9138820610dbbbfe76d1b393c419358`
- stderr: `544f49560e35888303f948649dfed154b43bc430e2463488b027098c24416304`

Decision: preserve the upstream-mirrored default-on wvSplitK route and rederive the local anchor/gap artifacts. Do not disable or narrow skinny GEMM merely to recover token 488. The issue remains open until a fresh implementer updates the committed evidence and manifest, the unchanged default `LoadedEngine::FromModelDir` test passes, a fresh reviewer mutates the default production route and artifact checks, and the operator reruns the gate.

**Comment by VikashLoomba** [comment]
The production-config K=10 per-prompt greedy capture is complete.

- prompt-tokenization mismatches: 0
- nondeterministic `(prompt, position)` cells across ten runs: 0/256
- local/current versus production oracle greedy differences: 8 cells, all prompt 7 positions 8 through 15
- prompt 10, including the disputed token 10 and resulting suffix, matches production oracle greedy exactly

Prompt 7 diverges at a tied first cell and then follows a different autoregressive suffix. The separate teacher-forced run on the exact local prompt-7 prefix reports a zero gap at every local suffix cell, so the intended refreshed gate result is 15/16 strict-greedy prompts plus one oracle-tied prompt, with 0.0-nat maximum teacher-forced gap and no forward divergence.

The production greedy files are deterministic and retained as:

- `greedy_ids.production.npy`: `c120eec482e2029037269f55721f9100c0f4717c956017aeaf51346abfd4b9c8`
- `greedy_dist.production.npy`: `1924f7a8915d367ea462975d0b5f91da78ba0f36d922c0d563abff9457ec16c8`
- result: `92460eb055228491c7e30c8bef83d7ce4a909db1c6c925ef31d0fba4fdf4c5d4`

The capture used `enforce_eager=False`; the engine resolved full-and-piecewise graph mode and default-on ROCm skinny GEMM. The committed capture scripts currently force eager mode, so the repair must make the production denominator reproducible rather than silently copying scratch artifacts produced under a mode the scripts cannot express. Eager may remain an explicitly named diagnostic arm, not the default acceptance denominator.

**Comment by VikashLoomba** [comment]
The diagnostic eager rerun isolates the remaining provenance question.

Running the currently committed `scripts/qwen3-oracle-capture.py` unchanged against the restored image produced `greedy_ids.npy` SHA-256 `865293612ec0e5da77a188c8513cca80f0efcb4235fae37b380530a0ca2413b3`, byte-identical to the committed historical file. The runtime log confirms `enforce_eager=True`, compilation disabled, and graph mode `NONE`.

Therefore the old oracle artifacts are reproducible; their difference from the new production capture is not explained by the restored dependency stack. It is the capture mode:

- historical/committed eager greedy: exact existing `greedy_ids.npy`
- current production greedy: deterministic K=10, `enforce_eager=False`, full-and-piecewise graph mode

The old script's eager-only constructor is now demonstrated to select a different oracle execution mode on this workload. It cannot remain the default reproduction path for a gate whose governing policy requires vLLM production configuration. The #2772 repair must make production mode the script default, retain eager only as an explicit diagnostic option, refresh all four ROCm oracle/anchor artifacts with matching provenance, and update the manifest and test comments accordingly.

Diagnostic evidence: `/tmp/qwen35-skinny-oracle-eager-k1-2772`; stdout SHA-256 `c1eedccd8f768ad8a8826b8903e25017fe93d06bb5b64b89a6fd4e454c1f2044`, stderr `c14cd374bc3f7bbb95e550f00aa605e7320f4d2d7ad7d82bbb8b239f3af98647`.
